code/modules/hydroponics/biogenerator.dm

```diff
@@ -1,337 +1,354 @@
-	anchored = 1
+	density = TRUE
+	anchored = TRUE
 	use_power = IDLE_POWER_USE
 	idle_power_usage = 40
-	var/processing = 0
-	var/obj/item/reagent_containers/glass/beaker = null
-	var/points = 0
-	var/menustat = "menu"
+	/// Is the biogenerator curretly grinding up plants?
+	var/processing = FALSE
+	/// The container that is used to store reagents from certain products.
+	var/obj/item/reagent_containers/glass/container
+	/// The amount of biomass stored in the machine.
+	var/biomass = NONE
+	/// Used to modify the cost of producing items. A higher number means cheaper costs.
 	var/efficiency = 0
+	/// Used to modify how much biomass is produced by grinding plants. A higher number means more biomass.
 	var/productivity = 0
-	var/max_items = 40
+	/// A list of plants currently stored plants in the biogenerator.
+	var/list/stored_plants = list()
+	/// The maximum amount of plants the biogenerator can store.
+	var/max_storable_plants = BASE_MAX_STORABLE_PLANTS
+	/// A reference to the biogenerator's research, which contains designs that it can build.
 	var/datum/research/files
-	var/list/show_categories = list("Food", "Botany Chemicals", "Organic Materials", "Leather and Cloth")
-	var/list/timesFiveCategories = list("Food", "Botany Chemicals", "Organic Materials")
+	/// A list which holds all categories and items the biogenator has available. Used with the UI to save having to rebuild this every time someone opens it.
+	var/list/product_list = list()
+	/// The categories in which the various items belong to.
+	var/list/categories = list("Food", "Botany Chemicals", "Organic Materials", "Leather and Cloth")
 
 /obj/machinery/biogenerator/Initialize(mapload)
 	. = ..()
-	files = new /datum/research/biogenerator(src)
-	create_reagents(1000)
+	files = new(src)
 	component_parts = list()
 	component_parts += new /obj/item/circuitboard/biogenerator(null)
 	component_parts += new /obj/item/stock_parts/matter_bin(null)
 	component_parts += new /obj/item/stock_parts/manipulator(null)
 	component_parts += new /obj/item/stack/sheet/glass(null)
 	component_parts += new /obj/item/stack/cable_coil(null, 1)
 	RefreshParts()
+	update_ui_product_list()
 
 /obj/machinery/biogenerator/Destroy()
-	QDEL_NULL(beaker)
+	QDEL_NULL(container)
 	QDEL_NULL(files)
+	QDEL_LIST(stored_plants)
 	return ..()
 
 /obj/machinery/biogenerator/ex_act(severity)
-	if(beaker)
-		beaker.ex_act(severity)
+	container?.ex_act(severity)
 	..()
 
 /obj/machinery/biogenerator/handle_atom_del(atom/A)
 	..()
-	if(A == beaker)
-		beaker = null
-		update_icon()
-		updateUsrDialog()
+	if(A != container)
+		return
+	container = null
+	update_icon()
+	SStgui.update_uis(src)
 
 /obj/machinery/biogenerator/RefreshParts()
-	var/E = 0
-	var/P = 0
-	var/max_storage = 40
+	efficiency = 0
+	productivity = 0
 	for(var/obj/item/stock_parts/matter_bin/B in component_parts)
-		P += B.rating
-		max_storage = 40 * B.rating
+		productivity += B.rating
+		max_storable_plants = BASE_MAX_STORABLE_PLANTS * B.rating
```

Comment:
This makes it *look* like the order of matter bin insertion matters (only the last one takes effect). However there is only one matter bin, so technically this is not a bug. But still confusing.

Consider setting `max_storage_plants` to zero - just like `efficiency` and `productivity` - and doing the `+=` here as well